source/agora/utils/Log.d

```diff
@@ -49,21 +49,24 @@ public struct Logger
         import core.memory;
         this.logger = Ocean.Log.lookup(moduleName);
         this.logger.buffer(new char[](16384));
-        GC.addRoot(cast(void*)this.logger);
-    }
-
-    /// dtor
-    public ~this ()
-    {
-        import core.memory;
-        GC.removeRoot(cast(void*)this.logger);
     }
 
     public void opDispatch (string call, Args...) (Args args)
     {
         try
         {
-            mixin("this.logger." ~ call ~ "(args);");
+            import core.memory : GC;
+            if (GC.inFinalizer())
+            {
+                writeln("allowing logging from the destructor on a GC thread would risk running into segfaults, please see issue #1128");
+                return;
+            }
+            else if (logger is null)
+            {
+                assert(0,"\nplease make sure you are declaring the \nmixin AddLogger!(); statement on top, followed by:\nstatic this{}; followed by:\nstatic ~this{};");
+                return;
```

Comment:
Return is not needed if you use `assert(0)` or `throw new Exception`, the compiler does basic flow analysis